examples/research_projects/zero-shot-distillation/distill_classifier.py

```diff
@@ -0,0 +1,320 @@
+    )
+
+
+@dataclass
+class DistillTrainingArguments(TrainingArguments):
+    output_dir: Optional[str] = field(
+        default=None,
+        metadata={"help": "The output directory where the model predictions and checkpoints will be written."},
+    )
+    per_device_train_batch_size: int = field(
+        default=32, metadata={"help": "Batch size per GPU/TPU core/CPU for training."}
+    )
+    per_device_eval_batch_size: int = field(
+        default=128, metadata={"help": "Batch size per GPU/TPU core/CPU for evaluation."}
+    )
+    num_train_epochs: float = field(default=1.0, metadata={"help": "Total number of training epochs to perform."})
+    do_train: bool = field(default=True, metadata={"help": "Whether to run training of student model."})
+    do_eval: bool = field(
+        default=True,
+        metadata={
+            "help": (
+                "Whether to evaluate the agreement of the final student predictions and the teacher predictions"
+                "after training."
+            )
+        },
+    )
+
+
+class DistillationTrainer(Trainer):
+    def compute_loss(self, model, inputs, return_outputs=False):
+        target_p = inputs["labels"]
+        outputs = model(inputs["input_ids"], attention_mask=inputs["attention_mask"])
+        logits = outputs[0]
+
+        loss = -torch.sum(target_p * logits.log_softmax(dim=-1), axis=-1).mean()
+
+        if return_outputs:
+            return loss, outputs
+
+        return loss
+
+
+def read_lines(path):
+    lines = []
+    with open(path, "r") as f:
+        for line in f:
+            line = line.strip()
+            if len(line) > 0:
+                lines.append(line)
+    return lines
+
+
+def get_premise_hypothesis_pairs(examples, class_names, hypothesis_template):
+    premises = []
+    hypotheses = []
+    for example in examples:
+        for name in class_names:
+            premises.append(example)
+            hypotheses.append(hypothesis_template.format(name))
+    return premises, hypotheses
+
+
+def get_entailment_id(config):
+    for label, ind in config.label2id.items():
+        if label.lower().startswith("entail"):
+            return ind
+    logging.warning("Could not identify entailment dimension from teacher config label2id. Setting to -1.")
+    return -1
+
+
+def get_teacher_predictions(
+    model_path: str,
+    examples: List[str],
+    class_names: List[str],
+    hypothesis_template: str,
+    batch_size: int,
+    temperature: float,
+    multi_class: bool,
+    use_fast_tokenizer: bool,
+    no_cuda: bool,
+):
+    """
+    Gets predictions by the same method as the zero-shot pipeline but with DataParallel & more efficient batching
+    """
+    model = AutoModelForSequenceClassification.from_pretrained(model_path)
+    model_config = model.config
+    if not no_cuda and torch.cuda.is_available():
+        model = nn.DataParallel(model)
+        batch_size *= len(model.device_ids)
```

Comment:
This should be more restrictive: if launched with torch.distributed.launch, I fear this part will try to use all the GPUs number of GPUs times (once for each process launched).